sql/sql_optimizer.cc

```diff
@@ -10162,16 +10162,29 @@ static bool internal_remove_eq_conds(THD *thd, Item *cond,
 
       if (new_item == NULL)
       {
+        bool replace_always_true = false;
         if (is_cond_or)
         {
-          Item* cond_true =
-            new (thd->mem_root) Item_int((longlong)1, 1); // Always true
-          (void)li.replace(cond_true);
-          should_fix_fields = true;
+          if (dynamic_cast<Item_int*>(item))
+          {
+            Item_int* item_int = dynamic_cast<Item_int*>(item);
+            replace_always_true = item_int->value ? true : false;
+          }
+          else if (dynamic_cast<Item_bool_func2*>(item))
+          {
+            Item_bool_func2* func = dynamic_cast<Item_bool_func2*>(item);
+            Item** args = func->arguments();
+            Item* left_item = args[0];
+            Item* right_item = args[1];
+            replace_always_true = left_item->eq(right_item, 0) ? true : false;
+          }
         }
-        else
-        {
-          li.remove();
+
+        if (replace_always_true)
```

Comment:
You should do useless clauses elimination according the comparision type. Such as. 

(A=A) AND others_clause. you can remove (A=A) because it is always true, and it will not affect the expression result.  for OR expression.  (A <>A) OR other_clauses, (A<>A) can removed due to (A<>A) is false, the OR expression result is the result of `other_clauses`.

Reply:
let me think about this